### PR TITLE
feat(forge): multi chain deployment

### PR DESCRIPTION
**Pull request by joshieDo**
follow-ups:
* library - linking (currently it just aborts if it the project requires it)
* Messy `stdout` during concurrent operations (simulation / broadcasting / verification)
* Unite both single and multi deployment flow once the above two are solved.

### CODE REVIEW

**Comment by gakonst** [comment]
Ref #2609 for multi chain verification

**Comment by gakonst** [comment]
@joshieDo FYI #2609 now merged

**Comment by joshieDo** [comment]
Still going to add simple tests, but opening the PR already for review.

File: cli/src/cmd/forge/script/broadcast.rs
Comment (mattsse) [inline cli/src/cmd/forge/script/broadcast.rs:352]: let's add what those are

File: cli/src/cmd/forge/script/providers.rs
Comment (mattsse) [inline cli/src/cmd/forge/script/providers.rs:11]: please add docs

File: cli/src/utils.rs
Comment (mattsse) [inline cli/src/utils.rs:230]: use consistent pattern

File: config/src/lib.rs
Comment (mattsse) [inline config/src/lib.rs:832]: we should be able to merge this with the `get_etherscan_config_with_chain` changes on master

File: evm/src/executor/inspector/cheatcodes/fork.rs
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/fork.rs:124]: let's make this an thiserror::Error type instead

**Review by mattsse — CHANGES_REQUESTED** [review]
great,

with multichain as a conditional deployment strategy, the existing functions got a bit bloated. Suggesting cleaner separations.

changes in evm/src/executor/backend/mod.rs and config/src/lib.rs can be extracted and merged separately

File: cli/src/utils.rs
Comment (gakonst) [reply cli/src/utils.rs:230]: should we log transactions separated by chain?

Comment (joshieDo) [reply cli/src/utils.rs]: yes definitely, it's part of the `stdout` follow-up i mentioned

File: cli/src/cmd/forge/script/broadcast.rs
Comment (joshieDo) [reply cli/src/cmd/forge/script/broadcast.rs]: ended up renaming it (`bundle_transactions`) and changing the doc

File: cli/src/cmd/forge/coverage.rs
Comment (joshieDo) [reply cli/src/cmd/forge/coverage.rs:273]: runner.known_contracts is already a `ContractsByArtifact`. I had to add `Clone` to `ContractsByArtifact`, which leads to an error if I leave as it was.

File: cli/tests/it/script.rs
Comment (joshieDo) [inline cli/tests/it/script.rs:403]: Fixing the test here since this **should** be failing, and it did get [fixed](https://github.com/joshieDo/foundry/blob/d85c83a3f604431cb886c5aab048fc35686ab303/cli/src/cmd/forge/script/mod.rs#L352-L353) in this branch.

File: cli/src/cmd/forge/script/broadcast.rs
Comment (joshieDo) [reply cli/src/cmd/forge/script/broadcast.rs:226]: Tried to make it clearer. Lmk if after this recente changes, you'd really separate it further... I feel like it's quite contained now. 

Eventually (when library linking is done), the goal is that the single deployment conditionals mostly disappear, since single deployment is a multi-deployment of one...

File: evm/src/executor/mod.rs
Comment (mattsse) [inline evm/src/executor/mod.rs:599]: `VecDeque<BroadcastableTransaction>` is used a lot,

does it make sense to make a new type for this as well?

File: cli/src/cmd/forge/script/executor.rs
Comment (mattsse) [inline cli/src/cmd/forge/script/executor.rs:275]: this could be a pitfall if `fork_url` ends up being None

should we use an error instead?

File: cli/src/cmd/forge/script/executor.rs
Comment (mattsse) [inline cli/src/cmd/forge/script/executor.rs:243]: also dangerous if None

File: cli/src/cmd/forge/script/executor.rs
Comment (mattsse) [inline cli/src/cmd/forge/script/executor.rs:131]: this could be removed if the futures return this instead?

**Comment by joshieDo** [comment]
added tests

File: cli/src/cmd/forge/script/broadcast.rs
Comment (mattsse) [inline cli/src/cmd/forge/script/broadcast.rs:347]:
```suggestion
            .wrap_err(|| {
```

File: cli/src/cmd/forge/script/broadcast.rs
Comment (mattsse) [inline cli/src/cmd/forge/script/broadcast.rs:380]: peekable is unused here?

File: cli/src/cmd/forge/script/broadcast.rs
Comment (mattsse) [inline cli/src/cmd/forge/script/broadcast.rs:384]: nit: switch order

File: cli/src/cmd/forge/script/cmd.rs
Comment (mattsse) [inline cli/src/cmd/forge/script/cmd.rs:70]: nit: we could add something like `EvmOpts::ensure_fork_url -> Result<&str, MissingForkErr>`

File: cli/src/cmd/forge/script/cmd.rs
Comment (mattsse) [inline cli/src/cmd/forge/script/cmd.rs:99]: `ScriptArgs::run_script` is now >200 lines with large if else branches.

can we separate the `if self.resume || (self.verify && !self.broadcast)`condition into separate functions?

File: cli/src/cmd/forge/script/cmd.rs
Comment (mattsse) [inline cli/src/cmd/forge/script/cmd.rs:136]: `if self.debug { } else {}` could be reduced to 

`if self.debug {...; return}`? to get rid of the else branch indent?

**Review by mattsse — CHANGES_REQUESTED** [review]
great progress.

a couple of smol nits, and I suggest we separate `ScriptArgs::run_script` because this is the main entry point and is currently a bit hard to read, especially for debugging.

Suggest we move the if else branches to individual functions, and add some traces to highlight what branch is called in `run_script`

File: cli/src/cmd/forge/script/broadcast.rs
Comment (joshieDo) [reply cli/src/cmd/forge/script/broadcast.rs:380]: it's used to check if the next rpc url is different. If so, it creates a `ScriptSequence` from all the collected transactions so far.

File: cli/src/cmd/forge/script/cmd.rs
Comment (mattsse) [inline cli/src/cmd/forge/script/cmd.rs:126]: this is a bit much can we use a helper type here.

since this function replaces the values it can just take ownership and return the modified values

File: cli/src/cmd/forge/script/executor.rs
Comment (mattsse) [inline cli/src/cmd/forge/script/executor.rs:160]: I'd rather move this hint to the `futs` directly:

`let futs: Vec<....> = ...`

File: evm/src/executor/inspector/cheatcodes/util.rs
Comment (mattsse) [inline evm/src/executor/inspector/cheatcodes/util.rs:34]:
```suggestion
    pub rpc: Option<RpcUrl>,
```

**Review by mattsse — CHANGES_REQUESTED** [review]
last smol nits

**Review by mattsse — APPROVED** [review]
I think this is ready now, @gakonst

**Comment by gakonst** [comment]
Would love if @aureliusbtc @mds1 or @hexonaut took it for a test before we merge

**Comment by mds1** [comment]
Planning to test this week! Will update here afterwards

**Comment by hexonaut** [comment]
We still haven't switched to Foundry deployment yet. It's one of the last things we are still using dapp tools for.

Will test these features when we switch over, but probably not for this PR.

**Comment by hexonaut** [comment]
Finally got around to testing this. With [this PR](https://github.com/makerdao/dss-bridge/pull/2) I'm seeing a non-descript error. Running: `forge script script/DeployExistingTokenBridge.s.sol:DeployExistingTokenBridge --use solc:0.8.14 --rpc-url $ETH_RPC_URL --broadcast --verify -vvvv`

Which produces:

```
The application panicked (crashed).
Message:  called `Option::unwrap()` on a `None` value
Location: evm/src/executor/backend/mod.rs:587
```

The crash appears to happen on this line when switching forks twice: https://github.com/makerdao/dss-bridge/pull/2/files#diff-e2a54d4d0ad2ae2494877491f14dffba4442e3ada78d1f1aecf7477e1857345aR47

**Comment by hexonaut** [comment]
As I said in our Discord chat, I updated to latest here and it seems to work fine now, so not sure what happened. Appears to work great now!

**Comment by gakonst** [comment]
@joshieDo could you resolve conflicts and feel free to merge! nice work.

**Comment by gakonst** [comment]
bump @joshieDo PTAL re: conflicts, would like to merge this week if poss

**Comment by joshieDo** [comment]
resolved pretty easily (maybe too easyily..). There were some changes which i was not aware on the `master` branch, and i'm not quite sure how that may impact this branch

waiting for the CI, and going to run some manual tests meanwhile

File: evm/src/executor/backend/mod.rs
Comment (joshieDo) [inline evm/src/executor/backend/mod.rs:860]: This was removed since it was causing trouble with https://github.com/foundry-rs/foundry/pull/2541/commits/3d09814c3ef478043793e21e40a2bf5e7b0ade88 .We assume that the caller is not part of the state when **first** selecting the fork. Caller should not be persistent, and this goes against that.


This was first introduced to tackle an error when reverting: https://github.com/foundry-rs/foundry/pull/2989.  However, by removing this block of code the associated test still passes.

**Comment by joshieDo** [comment]
manual tests of deployment + verification successfully passed. Pending CI i think we can merge, and tackle the `inevitable` issues as they come

**Comment by gakonst** [comment]
SHIP IT

**Comment by PatrickAlphaC** [comment]
So, what's the API look like on this?

https://ethereum.stackexchange.com/search?q=two+chains+foundry

How do I actually use this for a multi-chain deployment?

**Pull request merged**
